File: Codeforces/666B.cpp
#include <bits/stdc++.h>
using namespace std;

const int N=3009;

typedef pair<int,int> ii;

typedef vector<int> vi;
typedef vector<ii> vii;

int n, m, cost[N][N]={}, b[N]={}, optimal = 0, ans[5]={};
vi a[N];
vii v1[N], v2[N];

void doc(){
	scanf("%d%d",&n,&m);
	for(int i=0; i<m; i++){
		int x,y;
		scanf("%d%d",&x,&y);
		a[x].push_back(y);
	}
}

void BFS(int s, int d[]){
	queue<int> q;
	q.push(s);
	for(int i=1; i<=n; i++) d[i] = -1, b[i] = false;
	d[s] = 0, b[s] = true;
	while(!q.empty()){
		int u = q.front(); q.pop();
		for(int i=0; i<a[u].size(); i++){
			int v = a[u][i];
			if(!b[v]){
				b[v] = true;
				d[v] = d[u]+1;
				q.push(v);
			}
		}
	}
}

void update(int a, int b, int c, int d){
	optimal = cost[a][b] + cost[b][c] + cost[c][d];
	ans[0] = a, ans[1] = b, ans[2] = c, ans[3] = d;
}

void findPath(int b, int c){
	// if cannot reach c from b -> impossible
	if(cost[b][c]==-1) return;

	// loop through first 5 largest a and d
	// why 5? because duplicates of b and c
	for(int i=0; i<v2[b].size() && i<6; i++)
		for(int j=0; j<v1[c].size() && j<6; j++){
			int a = v2[b][i].second, d = v1[c][j].second;
			if(a==b || a==c || a==d || b==c || b==d || c==d) continue;
			if(optimal < cost[a][b] + cost[b][c] + cost[c][d]) update(a,b,c,d);
		}
}

void xuly(){
	// calculate the cost for every path
	for(int i=1; i<=n; i++) BFS(i,cost[i]);

	// v1 is normal graph, v2 is reversed graph
	for(int i=1; i<=n; i++){
		for(int j=1; j<=n; j++){
			if(cost[i][j]!=-1) v1[i].push_back(ii(cost[i][j],j));
			if(cost[j][i]!=-1) v2[i].push_back(ii(cost[j][i],j));
		}
		sort(v1[i].begin(), v1[i].end(), greater<ii>());
		sort(v2[i].begin(), v2[i].end(), greater<ii>());
	}

	// loop through every possible b and c
	for(int b=1; b<=n; b++)
		for(int c=1; c<=n; c++)
			findPath(b,c);

	// print answer
	for(int i=0; i<4; i++) printf("%d ",ans[i]);
	printf("\n");
}

int main(){
	doc();
	xuly();
	return 0;
}

File: Codeforces/666B.py
fo = open("test.out","w")

inp = input().split()
n,m = int(inp[0]),int(inp[1])
n+=1

inf = int(10E8)

vector = [[] for _ in range(n)]
front = [[] for _ in range(n)]
cost = [[inf for _ in range(n)] for __ in range(n)]
b = [False for _ in range(n)]

for i in range(n): cost[i][i] = 0

def DFS(root, u):
	b[u]=True
	for v in vector[u]:
		cost[root][v] = min(cost[root][u]+1, cost[root][v])
		if not b[v]: DFS(root,v)
		
def BFS(root):
	q = []
	q.insert(0,root)
	while q!=[]:
		u = q.pop()
		if b[u]: continue
		b[u] = True
		for v in vector[u]:
			cost[root][v] = min(cost[root][v], cost[root][u]+1)
			q.insert(0,v)

for i in range(m):
	inp = input().split()
	u,v = int(inp[0]),int(inp[1])
	cost[u][v] = 1
	vector[u].append(v)

# for i in range(n):
	# for j in range(n): b[j]=False
	# # DFS(i,i)
	# BFS(i)
	
for k in range(1,n):
	for i in range(1,n):
		for j in range(1,n):
			cost[i][j] = min(cost[i][j], cost[i][k]+cost[k][j])

ans = -inf
path = []

def findPath(s):
	global ans
	global path
	q = []
	q.append([0,s,0,[s]])
	while q != []:
		for i in q: fo.write(str(i)+'\n')
		current = q.pop(0)
		verNum, u, du = current[0], current[1], current[2]
		cpath=[]
		if verNum < 3:
			for v in range(1,n):
				if not(v in current[3]) and (cost[u][v]<inf):
					cpath = list(current[3])
					cpath.append(v)
					q.append([verNum+1, v, du+cost[u][v], cpath])
		elif ans <= du:
			ans, path = du, current[3]
			# print(ans, path)

for i in range(1,n):
	fo.write(str(i)+'\n')
	findPath(i)

# print('--')

for i in path:
	print(i, end=' ')
